### PR TITLE
fix(auth): Add criteria for resolving among ambiguous email addresses

### PR DESCRIPTION
**Pull request by RyanSkonnord**
Improve on how we choose from among multiple users who share an email address, using the strategy discussed in https://getsentry.atlassian.net/browse/ER-878. That is: first prefer verified email addresses, then prefer users who have a membership in the organization whose provider is being used.

### CODE REVIEW

**Comment by RyanSkonnord** [comment]
See https://github.com/getsentry/sentry/pull/30356 for old context.

See also https://github.com/getsentry/sentry/pull/30927 for a possible future direction where I get a little fancier.

**Comment by RyanSkonnord** [comment]
On hold until merging https://github.com/getsentry/sentry/pull/30947

File: src/sentry/auth/email.py
Comment (JoshFerge) [inline src/sentry/auth/email.py:89]: could you use `flat=True` on the values_list call so you don't have to unpack the tuple?

File: src/sentry/auth/email.py
Comment (JoshFerge) [inline src/sentry/auth/email.py:90]: is this line necessary given the query is filtered by the candidates?

File: src/sentry/auth/email.py
Comment (JoshFerge) [inline src/sentry/auth/email.py:99]: i'm not sure if this logic is correct. we are checking the UserEmail user's email against the parent email, but not that the UserEmail's email matches the UserEmail user's email. (say that 3 times fast!!). I think we need to check that the `UserEmail.is_primary() == True` as well here.

I think that this would not resolve correctly currently if there were a user with multiple secondary emails and we were attempting to resolve the primary.

I also wonder if UserEmail is case normalized

File: src/sentry/auth/email.py
Comment (RyanSkonnord) [reply src/sentry/auth/email.py:90]: The point is to return a subset of the original `UserEmail` objects. The query is returning `User` IDs.

File: src/sentry/auth/email.py
Comment (RyanSkonnord) [reply src/sentry/auth/email.py]: I want to say the logic checks out because `self.parent.email` is already guaranteed to be the UserEmail's email because that's what we queried for; that is, it's equivalent to writing `if ue.user.email == ue.email`. But you're right, it's much better to just call `if ue.is_primary()` (which -- hey! -- is also just `return self.user.email == self.email`).

As for being case-normalized, yeah, that looks messy. We seem to habitually write queries like `email__iexact` to treat it as case-insensitive, but `is_primary` doesn't respect that. It probably needs a fix, but one that would go on the `UserEmail` model class.

File: src/sentry/auth/email.py
Comment (JoshFerge) [reply src/sentry/auth/email.py:90]: 👍🏼

**Review by JoshFerge — APPROVED** [review]
nice stuff.

**Pull request merged**
